RFC 477: Policy Validation (#478)
aws/aws-cdk-rfcs: text/0477-policy-validation.md

```ts
// globally for the entire app (an app is a stage)
import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';

const app = new App({
  validationPlugins: [
    new CfnGuardValidator({
      rules: [
        // rules can be specified inline
        Rules.fromAsset('../my-local-rules'),
        
        // or from a remote location
        Rules.fromDownload('https://somelocation.com/company-rules', {
          auth: {},
        }),
        
        // or from an S3 location
        Rules.fromS3(s3Location),
        
        // or by calling an external endpoint
        Rules.fromRequest('POST', 'https://someendpoint.com/validate'),
      ],
      options: {
        someOption: "value",
      },
    }),
  ],
});

// only apply to a particular stage
const prodStage = new Stage(app, 'ProdStage', {
  validationPlugins: [...],
});
```

The details of how to configure the plugin will be specific to each plugin. For example, the `CfnGuardValidator` shown
above has a `rules` property which is an array of `Rules` objects. The `Rules` class is a helper class that can be used
to specify where the rules are located.

> CDK Policy Validation should never be used as a deployment gate or as the enforcement layer. It should instead be

Reviewer: Yeah we should add some more explanation here.  We're trying to say that you should not rely on this feature in order to prevent the deployment of misconfigured infrastructure. It's fine if it runs as part of a deployment pipeline and fails the pipeline if it is not compliant, but it should be viewed as more of "fail fast(er)" tool rather than a preventative measure. I guess both could be viewed as a type of "gate" though.

> CDK Policy Validation should never be used as a deployment gate or as the enforcement layer. It should instead be

Reviewer: I'm changing this to `an enforcement layer` I don't think it should be thought of as an enforcement layer at all.

> CDK Policy Validation should never be used as a deployment gate or as the enforcement layer. It should instead be

Reviewer: I've added a note.

the reason for the suppression. Each plugin is responsible for handling it's own
suppressions and exposes a standard API.

A suppression can be added for all resources under a construct scope. For
example to add suppressions for an entire stack.

```ts
import { Stack } from 'aws-cdk-lib';
import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';

const stack = new Stack(app, 'DevStack');
CfnGuardValidator.addSuppression(stack, {
  rule: 'S3BucketEncryption'
  reason: 'Dev environment buckets do not have to be encrypted'
});
```

They can also be added for specific resources.

```ts
import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';

class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    const bucket = new Bucket(this, 'MyBucket');
    CfnGuardValidator.addSuppression(bucket, {
      rule: 'S3BucketEncryption',
      reason: 'This bucket does not require encryption because xyz',
    });
  }
}
```

Each plugin will determine how the suppressions are handled for the specific tool used, for example a plugin might

> add to the resource metadata.

Reviewer: yeah it would then read it during validation. That way you could have a single guard rule, for example

```
let s3_bucket_public_read_prohibited = Resources.*[ Type == 'AWS::S3::Bucket'
  Metadata.guard.SuppressedRules not exists or
  Metadata.guard.SuppressedRules.* != "S3_BUCKET_PUBLIC_READ_PROHIBITED"
]

rule S3_BUCKET_PUBLIC_READ_PROHIBITED when %s3_bucket_public_read_prohibited !empty {
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration exists
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.BlockPublicAcls == true
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.BlockPublicPolicy == true
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.IgnorePublicAcls == true
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.RestrictPublicBuckets == true
  <<
    Violation: S3 Bucket Public Write Access controls need to be restricted.
    Fix: Set S3 Bucket PublicAccessBlockConfiguration properties for BlockPublicAcls, BlockPublicPolicy, IgnorePublicAcls, RestrictPublicBuckets parameters to true.
  >>
}
```

This rule could be evaluated by running guard as part of a guard plugin _or_ it could be run in a CFN Hook when the cloudformation template is actually deployed. The suppression needs to be persisted in a way that allows it to be evaluated during deployment as well, so rather than create a separate suppression mechanism that only works for CDK plugins, we're trying to create one that works for both.

> add to the resource metadata.

Reviewer 2: What you are describing is worse than "the plugin takes care of it" though. This is "the user-written ruleset must take suppressions into account".

I wonder if these aren't in actuality two different features:

* One to actually suppress findings (I think this might be good to have anyway, like how people would like suppressions for our warnings, which we currently don't offer)
* One to smuggle out-of-band data from the CDK app to the rule set, so that the rule set can use it in its rules

> add to the resource metadata.

Reviewer: > What you are describing is worse than "the plugin takes care of it" though. This is "the user-written ruleset must take suppressions into account".

I agree it's not great, but what are the alternatives? Using the CFN guard example above, the _only_ option for suppressions is to have the suppression somewhere in the template. Guard doesn't have the ability to read data from somewhere else. So if we wrote some suppression logic into the plugin it would allow the developer to get passed local development, but then it would fail at deployment because the deployment system doesn't know about the suppression you added.

The only other alternative I can think of is to also build a feature into CFN hooks around suppressions. CFN hooks would read suppression info from somewhere (not sure it can be given access to anything other than the template though).

> add to the resource metadata.

Reviewer: After talking to Rico I split this section into two sections, 1 for each of suppressions and exemptions. @rix0rrr let me know what you think.

See [Appendix A - High level design](https://quip-amazon.com/lDPqAisXjVP6#temp:C:CaFafef0ee17e804c3bb00434ff5).

### Is this a breaking change?

No.

### What alternative solutions did you consider?

We have considered two alternative solutions:

#### Validation in the CLI

With this solution, the validation would be done in the CLI, rather than in the framework. With this approach, users
would configure the validation plugins using a config file (cdk.json would be a natural candidate):

```json
"validations": {
    "package": "cdk-cfn-guard",
    "class": "CfnFuardValidation",
    "version": "1.2.3"
}
```

The CLI would read this file, load the appropriate plugins either from the filesystem or directly from NPM, and call
them to get a validation report.

We eventually discarded this option for the following reasons:

* All plugins would have to be written in TypeScript/JavaScript for them to be consumable by the CLI.
* The framework has access to the construct tree in memory, with all the contextual information that is needed to
  provide a better error message, such as stack traces. This is useful to point out, for example, in which line 

>   an offending construct was defined. The CLI, on the other hand, only has access to the synthesized template.

Reviewer 3: Not accurate, the CLI also has access to the `manifest.json` file, which can contain metadata produced by the framework. Its just easier to access this information from the framework, but possible via the CLI as well.

Author: Fair enough. Item removed.

CDK Policy Validation should never be used as a deployment gate or as the enforcement layer. It should instead be
thought of as developer enablement. This is not a tool that organizations should use to ensure developers actually
configure the validations. What prevents them from simply omitting it, or making a mistake? Well, nothing really, when
developers are writing applications, there's always going to be some setup they need to do. The goal of CDK Policy
Validation is to minimize it, and make it as easy as possible. For example, a typical workflow could be:

* Developer writes a CDK application without the correct validations config.
* Developer deploys non-compliant stacks.
* Deployment guardrails catch these violations, and instruct the developer to add a validations property to their
  application.
* Developer adds the validations property, and avoids these violations going forward.

Policy as code tools are integrated with CDK through a plugin mechanism so in order to add policy enforcement for a
specific tool you need to first specify the plugin to use. This can be done by importing the plugin and applying it to a

> scope.

Reviewer: yeah it should be "stage"

```ts
// globally for the entire app (an app is a stage)
import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';

const app = new App({
  validationPlugins: [
    new CfnGuardValidator({
      rules: [
        Rules.fromAsset('../my-local-rules'),
        Rules.fromDownload('https://somelocation.com', {
          auth: {},
        }),
        Rules.fromS3(s3Location),
        Rules.fromRequest('https://someendpoint.com'),
        ],
      options: {
        someOption: "value",
      },
    }),
  ],
});

// only apply to a particular stage
const prodStage = new Stage(app, 'ProdStage', {
  validationPlugins: [...],
});
```

#### Alternative

```ts
const app = new App();

> Validations.of(app).add(new CfnGuardValidator());

Reviewer: Done.

    > Add to construct properties for `cdk-app/MyStack/Bucket`
      `encryption: BucketEncryption.KMS`

  
Validation failed. See above reports for details
```

### Adding Exemptions/Suppressions

> Prior art: [cdk-nag](https://github.com/cdklabs/cdk-nag).

There may be cases where you would like to suppress a certain rule. To do so you must specify the rule to suppress and
the reason for the suppression.

A suppression can be added for all resources under a construct scope. For
example to add suppressions for an entire stack.

```ts
import { Stack, Validations } from 'aws-cdk-lib';

const stack = new Stack(app, 'DevStack');
Validations.of(stack).addSuppression({
  rule: 'S3BucketEncryption'
  reason: 'Dev environment buckets do not have to be encrypted'
});
```

They can also be added for specific resources.

```ts
class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    const bucket = new Bucket(this, 'MyBucket');
    Validations.of(bucket).addSuppression({
      rule: 'S3BucketEncryption',
      reason: 'This bucket does not require encryption because xyz',
    });
  }
}
```

> Each plugin will determine how the suppressions are handled for the specific tool used, but the CDK will also synthesize

Reviewer: I think the plugins have to access suppressions via the template. The rules that are run locally should be the same rules that are eventually run as part of CFN hooks so it needs to be the rules themselves that have the logic for identifying suppressions and not the plugin.

I think the suppressions need to be written to each individual resource. Maybe instead of the metadata we should use tags instead since those will be available in the template as well as the resource itself (and available to Config). 

Since the plugins run after the template is synthesized I don't think we can have the plugin report back anything since it is too late to modify the template at this point.

> And again going back to my previous question, how will plugins know that a specific suppression relates to them and not to some other plugin?

I don't think this matters. If there are multiple plugins that both have the same rule I think a single suppression would be fine.

the suppressions into the template metadata.

```json
{
  "Resources": {
    "MyBucket": {
      "Type": "AWS::S3::Bucket",
      "Properties": {...},
      "Metadata": {
        "cdk_validations": {
          "rules_to_suppress": [
            "rule": "S3BucketEncryption",
            "reason": 'This bucket does not require encryption because xyz'
          ]
        }
      }
    }
  }
}
```

This behavior can be disabled by providing the `validationMetadata: false` property.

```ts
new App({
  validationMetadata: false,
});
```

> **Example of how CFN Guard rules handle suppression:**

Author: Removed.

<https://github.com/aws-cloudformation/aws-guard-rules-registry/blob/main/rules/aws/amazon_s3/s3_bucket_server_side_encryption_enabled.guard>

Rule snippet:

```text
let s3_buckets_server_side_encryption = Resources.*[ Type == 'AWS::S3::Bucket'
  Metadata.guard.SuppressedRules not exists or
  Metadata.guard.SuppressedRules.* != "S3_BUCKET_SERVER_SIDE_ENCRYPTION_ENABLED"
]
```

The CFN Guard plugin would then generate that metadata in the template

```json
{
  "Resources": {
    "MyBucket": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        ...
      },
      "Metadata": {

>         "guard": {

Reviewer: This is changed so that now we will not write any metadata ourselves. Only the plugin will.

worst, may expose the customer to attacks or lawsuits.

By integrating policy-as-code tools in the synthesis flow, we give the confidence that these customers need to adopt the
CDK at a larger scale.

### What is the technical solution (design) of this feature?

See [Appendix A - High level design](https://quip-amazon.com/lDPqAisXjVP6#temp:C:CaFafef0ee17e804c3bb00434ff5).

### Is this a breaking change?

No.

### What alternative solutions did you consider?

We have considered two alternative solutions:

#### Validation in the CLI

With this solution, the validation would be done in the CLI, rather than in the framework. With this approach, users
would configure the validation plugins using a config file (cdk.json would be a natural candidate):

```json
"validations": {
    "package": "cdk-cfn-guard",
    "class": "CfnFuardValidation",
    "version": "1.2.3"
}
```

The CLI would read this file, load the appropriate plugins either from the filesystem or directly from NPM, and call
them to get a validation report.

We eventually discarded this option for the following reasons:

* All plugins would have to be written in TypeScript/JavaScript for them to be consumable by the CLI.

> * The CLI lacks contextual information that would be useful for tracing the problem back to where it was defined.

Author: Done.

* If an application is synthesized without using the CDK CLI, it’s not subject to policy validation.

#### CloudFormation hooks integration

Instead of integrating with each policy tool individually, we have considered the option of developing a single
mechanism, that would integrate with CloudFormation hooks. The idea was to invoke the Lambda function underlying a
CloudFormation hook and interpret its results in a similar way CloudFormation, that is, by failing synthesis if the
function reported a validation failure.

However, the Lambda functions invoked by CloudFormation hooks are not deployed to the customer’s account, but to an
account owned by AWS. This makes this solution not feasible.

> ### What are the drawbacks of this solution?

Reviewer 3: Another drawback I'd like to point out:

- If an organization wants validations to apply to all languages in the company, they have to setup a jsii publishing pipeline. But if the validation is invoked from the CLI, than all that is needed is one node package.

application that uses a given plugin may not have the necessary tools installed beforehand, when they synthesize the app
for the first time. Plugin authors must then make sure that developers can install these tools with as little friction
as possible to deliver a good user experience.

### What is the high-level project plan?

The first release of this feature will include integration with CloudFormation Guard. To make sure that our design is
not tied to a specific vendor, we will also work on a prototype for the integration with a second policy checking tool (
TBD). As usual with new CDK features, policy checking at synth time will be marked as experimental while we collect
feedback from the community, fix bugs and stabilize the API. Once it’s promoted to stable, we’ll go back to the
integration prototype and release it as a second officially supported integration.

### Are there any open issues that need to be addressed later?

No.

## Appendix

### Appendix A - High level design

![Class diagram for policy validation](https://github.com/aws/aws-cdk-rfcs/blob/otaviom/rfc-0477/images/policy-validation/validation.png?raw=true)

> The main abstraction of the model is that of a validation plugin. A plugin encapsulates a piece of logic that includes

Reviewer 3: Who installs the plugin itself? The framework or the user?

Reviewer 3: Yeap

Author: That should be the responsibility of the plugin.

above has a `rules` property which is an array of `Rules` objects. The `Rules` class is a helper class that can be used
to specify where the rules are located.

CDK Policy Validation should never be used as a deployment gate or as an enforcement layer. It should instead be
thought of as developer enablement. Ultimately, the development environment is under the control of individual 
developers and development teams. It's up to them to opt in to this feature, and to ensure that the policy 
validation is applied. The goal of CDK Policy Validation is to minimize the amount of set up needed during development, and make it as easy
as possible. For example, a typical workflow could be:

> A note on the term "deployment gate" above. It is perfectly acceptable to have
> policy at synth run as part of you CI or CD pipeline and prevent deployments.
> What we want to ensure is that users do not think of that "deployment gate" as
> a way to _ensure_ that their infrastructure deployment is compliant. It should
> instead be thought of as a way to "fail fast" and report issues before

> > deployment.

Reviewer 2: I know what you're trying to say, but I would say it differently. How about something a long the lines of:

*The validation performed by the CDK at synth time can be bypassed by developers, and can therefore not be relied on as the sole mechanism of validation in large organizations. Some other mechanism to validate the same rule set more authoritatively should be set up independently, like CloudFormation Hooks or AWS Config. Nevertheless, CDK's ability to evaluate the rule set during development is still useful as it will improve detection speed and developer productivity.*

Reviewer: ooh I like this better

* Developer writes a CDK application without the correct validations config.
* Developer deploys non-compliant stacks.
* Deployment guardrails catch these violations, and instruct the developer to add a validations property to their
  application.

> * Developer adds the validations property, and avoids these violations going forward.

Reviewer 2: I think this is covered? Whaddayatrynasay?

-------------------------------------

(Summary)

╔═══════════╤════════════════════════╗
║ Status    │ failure                ║
╟───────────┼────────────────────────╢
║ Plugin    │ CfnGuardValidator      ║
╚═══════════╧════════════════════════╝

(Violations)

Ensure S3 Buckets are encrypted with a KMS CMK (2 occurrences)

  Occurrences:
  
    - Construct Path: MyStack/MyCustomL3Construct/Bucket
    - Stack Template Path: ./cdk.out/MyStack.template.json
    - Creation Stack:
        └──  Bucket (MyStack/MyCustomL3Construct/Bucket)
             │ Library: aws-cdk-lib/aws-s3.Bucket
             │ Library Version: 2.50.0
             │ Location: new Bucket (/home/hallcor/tmp/cdk-tmp-app/node_modules/aws-cdk-lib/aws-s3/lib/bucket.js:1:12889)
             └──  MyCustomL3Construct (MyStack/MyCustomL3Construct)
                  │ Library: N/A - (Local Construct)
                  │ Library Version: N/A
                  │ Location: new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/custom-construct.ts:9:20)
                  └──  MyStack (MyStack)
                       │ Library: aws-cdk-lib.Stack
                       │ Library Version: 2.50.0

>                        │ Location: new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)

Reviewer 2: This tree is the wrong way round :) 

We can iterate on this a little. I love the idea but not the specifics of this particular rendering.

Reviewer: I actually did it this direction on purpose since it shows you the violation resource first, but I'm not tied to it. The `Location` info is also a little weird, but I'm not sure what the best option is for it. For example

```
new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/custom-construct.ts:9:20)
```

This file location is actually the location of `new s3.Bucket`

Reviewer: I've been convinced that the order should go top down. Updated.

Reviewer: I was convinced by the argument that we should go from most familiar to the user to least familiar. At the top they see something they are familiar with and can easily trace down until they see things they are not familiar with. Bottom up they start with something they might not even know exists and then have to dig down the trace until they hit something they recognize.

If we take this example, the user probably did something like

```ts
class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
      new MyCustomL3Construct(this, 'MyCustomL3Construct');
  }
}
```

They would recognize the stack and the L3 construct that they created. They might not even know that the L3 construct creates an S3 bucket. If we show the Bucket trace first they will have to dig down the trace until the reach the L3 construct which they recognize.

    - Resource Name: my-bucket
    - Locations:
      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm

  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
  How to fix:

>     > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`

Reviewer 2: Oh my God can we generate the correct code here? That'd be neat!

    > Add to construct properties for `cdk-app/MyStack/Bucket`

>       `encryption: BucketEncryption.KMS`

Reviewer 2: I like this, but this needs be propped further up front that there will be hints attached for fixing violations, and that's one of the value adds of this feature!

  
Validation failed. See above reports for details
```

### Adding Exemptions/Suppressions

> Prior art: [cdk-nag](https://github.com/cdklabs/cdk-nag).

There may be cases where you would like to suppress a certain rule. To do so you must specify the rule to suppress and

> the reason for the suppression. Each plugin is responsible for handling it's own

Reviewer 2: ```suggestion
the reason for the suppression. Each plugin is responsible for handling its own
```

suppressions and exposes a standard API.

A suppression can be added for all resources under a construct scope. For
example to add suppressions for an entire stack.

```ts
import { Stack } from 'aws-cdk-lib';
import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';

const stack = new Stack(app, 'DevStack');
CfnGuardValidator.addSuppression(stack, {
  rule: 'S3BucketEncryption'
  reason: 'Dev environment buckets do not have to be encrypted'
});
```

They can also be added for specific resources.

```ts
import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';

class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    const bucket = new Bucket(this, 'MyBucket');
    CfnGuardValidator.addSuppression(bucket, {
      rule: 'S3BucketEncryption',
      reason: 'This bucket does not require encryption because xyz',
    });
  }
}
```

Each plugin will determine how the suppressions are handled for the specific tool used, for example a plugin might
add to the resource metadata.

```ts
class CfnGuardValidator implements IValidatorPlugin {

>   public static addSuppression(scope: Construct, props: SuppressionProps) {

Reviewer 2: * `suppress()` ?
* `ignore()` ?

Reviewer: went with `suppress`

above has a `rules` property which is an array of `Rules` objects. The `Rules` class is a helper class that can be used
to specify where the rules are located.

The validation performed by the CDK at synth time can be bypassed by
developers, and can therefore not be relied on as the sole mechanism
of validation in large organizations. Some other mechanism to validate
the same rule set more authoritatively should be set up independently,
like CloudFormation Hooks or AWS Config. Nevertheless, CDK's ability
to evaluate the rule set during development is still useful as it will
improve detection speed and developer productivity.

The goal of CDK Policy Validation is to minimize the amount of set up needed
during development, and make it as easy as possible. For example, a typical
workflow could be:

* Developer writes a CDK application without the correct validations config.
* Developer deploys non-compliant stacks.
* Deployment guardrails catch these violations, and instruct the developer to add a validations property to their
  application.

#### Alternative

```ts
const app = new App();
Policy.of(app).add(new CfnGuardValidator());
```

> ### Validation Report

Author: Yes. Added.

                  └──  Bucket (MyStack/MyCustomL3Construct/Bucket)
                       │ Library: aws-cdk-lib/aws-s3.Bucket
                       │ Library Version: 2.50.0
                       │ Location: new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)
    - Resource Name: my-bucket
    - Locations:
      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm

  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
  How to fix:
    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
    > Add to construct properties for `cdk-app/MyStack/Bucket`
      `encryption: BucketEncryption.KMS`

  
Validation failed. See above reports for details
```

### Plugins

The CDK core framework is responsible for registering and invoking plugins and
then displaying the formatted validation report. The responsibility of the
plugin is to act as the translation layer between the CDK framework and the
policy validation tool. Responsibilities of the plugin may include things like:

- Bundling or installing the policy tool (cfn-guard, opa, etc)

> - Invoking the policy tool, parsing the output, and returning a

Reviewer 3: ```suggestion
- Bundling or fetching the rules to evaluate. 
- Invoking the policy tool, parsing the output, and returning a
```

                  │ Library Version: N/A
                  │ Location: new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:15:20)
                  └──  Bucket (MyStack/MyCustomL3Construct/Bucket)
                       │ Library: aws-cdk-lib/aws-s3.Bucket
                       │ Library Version: 2.50.0
                       │ Location: new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)
    - Resource Name: my-bucket
    - Locations:
      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm

  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
  How to fix:
    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
    > Add to construct properties for `cdk-app/MyStack/Bucket`
      `encryption: BucketEncryption.KMS`

  
Validation failed. See above reports for details
```

By default, the report will be printed in a human readable format. If you want a
report in JSON format, enable it using the `@aws-cdk/core:validationReportJson`
via the CLI or passing it directly to the application:

```ts
const app = new App({ 

>   context: { '@aws-cdk/core:validationReportJson': true }, 

Reviewer 3: This is a bit strange. Normally JSON output is needed in order for a machine to parse the output, but running `cdk synth` also produces additional stuff that will make the output non parsable. That is, it will contains some lines, then a JSON report, and some more lines? 

I think its going to be hard to make sure the CLI doesn't print anything when this feature is turned on, so I would suggest allowing to pass a file where the JSON report will be written to. Is this possible via context?

Reviewer: If the report returns anything that means it fails and synth will not write anything to `stdout`. The only thing written to `stdout` will be the report.

>   context: { '@aws-cdk/core:validationReportJson': true }, 

Reviewer: Just trying to stay consistent and match the feature flags https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.cx_api-readme.html.

the plugin in each CDK language.

#### Developing Plugins

If you need to develop your own policy validation plugin, either because one
does not exist for your policy tool or because and existing plugin does not meet
your use case, you start by creating a class that implements the
`IValidationPlugin` interface from `aws-cdk-lib`.

```ts
import { ValidationPluginReport } from './report';

/**
 * Represents a validation plugin that will be executed during synthesis
 */
export interface IValidationPlugin {
  /**
   * The name of the plugin that will be displayed in the validation
   * report
   */
  readonly name: string;

  /**
   * The method that will be called by the CDK framework to perform
   * validations. This is where the plugin will evaluate the CloudFormation
   * templates for compliance and report and violations
   */
  validate(context: IValidationContext): ValidationPluginReport;
}

/**
 * Context available to the validation plugin
 */
export interface IValidationContext {
  /**
   * The absolute path of all templates to be processed
   */
  readonly templatePaths: string[];
}

```

The `validate` method returns an instance of `ValidationPluginReport`, which

> tells the CDK whether the template is compliant, which violations were found

Reviewer 3: ```suggestion
tells the CDK whether the templates are compliant, which violations were found
```

(if any), and any metadata about the report. These are the report related 

> interfaces:

Reviewer 3: nit: this should be in the appendix or design sections, not the working backwards.
